receiver/kafkareceiver/consumer_franz.go

```diff
@@ -0,0 +1,477 @@
+					attribute.String(attrPartition, strconv.Itoa(int(partition))),
+				),
+			}
+			partitionConsumer.ctx, partitionConsumer.cancel = context.WithCancelCause(ctx)
+			c.assignments[topicPartition{topic: topic, partition: partition}] = &partitionConsumer
+		}
+	}
+}
+
+// lost must be set both on kgo.OnPartitionsLost and kgo.OnPartitionsReassigned
+// callbacks. Ensures that partitions that are lost (see kgo.OnPartitionsLost
+// for more details) or reassigned (see kgo.OnPartitionsReassigned for more
+// details) have their partition consumer stopped.
+// This callback must finish within the re-balance timeout.
+func (c *franzConsumer) lost(ctx context.Context, _ *kgo.Client,
+	lost map[string][]int32, fatal bool,
+) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	var wg sync.WaitGroup
+	for topic, partitions := range lost {
+		for _, partition := range partitions {
+			tp := topicPartition{topic: topic, partition: partition}
+			pc := c.assignments[tp]
+			delete(c.assignments, tp)
+			pc.cancel(errors.New(
+				"stopping processing: partition reassigned or lost",
+			))
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				pc.wg.Wait()
+			}()
+			c.telemetryBuilder.KafkaReceiverPartitionClose.Add(context.Background(),
+				1, metric.WithAttributes(attribute.String(attrInstanceName, c.id.Name())),
+			)
+		}
+	}
+	if fatal {
+		return
+	}
+	// Wait for all partition consumers to exit before committing marked offsets.
+	wg.Wait()
+	// NOTE(marclop) commit the marked offsets when the partition is rebalanced
+	// away from this consumer.
+	if err := c.client.CommitMarkedOffsets(ctx); err != nil {
+		c.settings.Logger.Error("failed to commit marked offsets", zap.Error(err))
+	}
+}
+
+// handleMessage is called on a per-partition basis.
+func (c *franzConsumer) handleMessage(pc *pc, msg kafkaMessage) error {
+	if pc.backOff != nil {
+		defer pc.backOff.Reset()
+	}
+
+	for {
+		err := c.consumeMessage(pc.ctx, msg, pc.attrs)
+		if err == nil {
+			return nil // Successfully processed.
+		}
+		// In the future, with Consumer Share Groups, messages not processed
+		// within a configurable timeout, are re-delivered to the consumer in
+		// the Share group, however, at the time of writing this feature isn't
+		// yet GA nor widely deployed.
+		// Share groups are generally more in line with observability and OTel
+		// collector use cases than traditional consumer groups.
+		// https://cwiki.apache.org/confluence/display/KAFKA/KIP-932%3A+Queues+for+Kafka.
+		// One possible exception is if the OTel collector is used for analytics
+		// pipelines, where it may make sense to make share groups opt-in.
+		if pc.backOff != nil && !consumererror.IsPermanent(err) {
+			backOffDelay := pc.backOff.NextBackOff()
+			if backOffDelay != backoff.Stop {
+				pc.logger.Info("Backing off due to error from the next consumer.",
+					zap.Error(err),
+					zap.Duration("delay", backOffDelay),
+				)
+				select {
+				case <-pc.ctx.Done():
+					return nil
```

Comment:
Could this cause the offset to be committed? i.e. if the partition is lost, but the receiver is not shutting down, then the partition context will be cancelled, but the main consumer context will not be cancelled, so the commit will go through.